pkg/host/numa.go

```diff
@@ -0,0 +1,313 @@
+	if err != nil {
+		return err
+	}
+
+	log.V(1).Info("Numa aware and hugepage memory tuning", "requiredMemory", requiredMemory, "domain", domain.Name)
+	if ok, err := n.isEnoughFreePagesAllocated(requiredMemory); !ok || err != nil {
+		if !ok {
+			return fmt.Errorf("not enough  memory in  preallocated amount of hugepages ")
+		}
+		return fmt.Errorf("unable to get free pages info: %w", err)
+	}
+
+	allocatableMemoryPerNode, err := n.getFreeNumaNodes(log, requiredMemory)
+	if err != nil {
+		return err
+	}
+
+	nodeCPUs := n.CPUTopology
+	for node, cpus := range nodeCPUs {
+		//	CPU cores with less pinned vm count go first
+		sort.SliceStable(cpus, func(i, j int) bool {
+			return pins[cpus[i]] < pins[cpus[j]]
+		})
+		nodeCPUs[node] = cpus
+	}
+
+	var totalMemory uint64
+	nodeIDs := make([]int, 0, len(allocatableMemoryPerNode))
+	for node, mem := range allocatableMemoryPerNode {
+		totalMemory += mem
+		nodeIDs = append(nodeIDs, node)
+	}
+
+	// nodes with more free memory go first
+	sort.SliceStable(nodeIDs, func(i, j int) bool {
+		return allocatableMemoryPerNode[i] > allocatableMemoryPerNode[j]
+	})
+
+	cpuCount := domain.VCPU.Value
+	numaCfg := &libvirtxml.DomainNuma{}
+	memNodesCfg := make([]libvirtxml.DomainNUMATuneMemNode, 0, len(allocatableMemoryPerNode))
+	vcpuPins := make([]libvirtxml.DomainCPUTuneVCPUPin, 0, cpuCount)
+
+	var (
+		cellID     uint
+		pinnedCPUs uint
+		nodeSet    []string
+		i          int
+	)
+
+	for _, nodeID := range nodeIDs {
+		mem := allocatableMemoryPerNode[nodeID]
+		var CPUs string
+
+		if pinnedCPUs < cpuCount {
+			k := float64(uint64(cpuCount)*mem) / float64(totalMemory)
+			s := uint(math.Ceil(k))
+
+			if cpuCount < pinnedCPUs+s {
+				s = cpuCount - pinnedCPUs
+			}
+			pinnedCPUs += s
+			selectedCPUS := nodeCPUs[nodeID][:s]
+
+			for _, c := range selectedCPUS {
+				vcpuPins = append(vcpuPins, libvirtxml.DomainCPUTuneVCPUPin{VCPU: uint(i), CPUSet: strconv.Itoa(c)})
+				CPUs += strconv.Itoa(i) + ","
+				i += 1
+			}
+			CPUs = strings.TrimRight(CPUs, ",") //nolint:staticcheck
+		}
+
+		nodeStr := strconv.Itoa(nodeID)
+
+		tmpCellID := cellID
+		cell := libvirtxml.DomainCell{
+			ID:     &tmpCellID,
+			Memory: uint(mem),
+			Unit:   KiB,
+		}
+		if CPUs != "" {
+			cell.CPUs = CPUs
+		}
+
+		nodeSet = append(nodeSet, nodeStr)
+		numaCfg.Cell = append(numaCfg.Cell, cell)
+		memNodesCfg = append(memNodesCfg, libvirtxml.DomainNUMATuneMemNode{
+			CellID:  tmpCellID,
+			Mode:    "strict",
+			Nodeset: nodeStr,
+		})
+
+		cellID += 1
+	}
+
+	numaTuneCfg := &libvirtxml.DomainNUMATune{
+		Memory: &libvirtxml.DomainNUMATuneMemory{
+			Mode:    "strict",
+			Nodeset: strings.Join(nodeSet, ","),
+		},
+		MemNodes: memNodesCfg,
+	}
+	domain.NUMATune = numaTuneCfg
+	if domain.CPU == nil {
+		domain.CPU = &libvirtxml.DomainCPU{}
+	}
+	domain.CPU.Numa = numaCfg
+
+	if domain.CPUTune == nil {
+		domain.CPUTune = &libvirtxml.DomainCPUTune{}
+	}
+
+	domain.CPUTune.VCPUPin = vcpuPins
+
+	return nil
+}
```

Comment:
`Tune` is called concurrently. At the moment the calculation can be wrong therefore.

Reply:
it is solving in other thread